policies: add GeoIP policy

gergosimonyi opened the pull request:
Closes #4433 

Introduces the new GeoIP policy with the functionality to match client ASN or country (for now).

~Opening a draft of this for a high-level overview. There are still many details to iron out like~

-   [x] Back-end tests
-   [x] Handle disabled geoip
-   [x] Handle IP missing in geoip database
-   [x] Handle empty lists of ASNs and countries
-   [x] Translations
-   [x] Types on `web`
-   [x] Multiselect on `web`

---

## Checklist

-   [x] Local tests pass (`ak test authentik/`)
-   [x] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [x] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [x] The code has been formatted (`make web`)

If applicable

-   [x] The documentation has been updated
-   [x] The documentation has been formatted (`make website`)

gergosimonyi committed:
add GeoIP policy

gergosimonyi commented on authentik/policies/geoip/countries.py:
```diff
@@ -0,0 +1,253 @@
+"""List of all ISO-3166-1 alpha-2 country codes"""
```
An alternative to this adding a dependency like [iso3166](https://github.com/deactivated/python-iso3166) or [django-countries](https://github.com/SmileyChris/django-countries).

Since ISO 3166 changes _infrequently_, I erred on the side of fewer dependencies.

BeryJu replied:
I don't mind adding a dependency for this, especially since we should add gettext calls to all these humanized names, and some dependency might have full translation for all the country names already

gergosimonyi replied:
Added `django-countries`. I'm unfamiliar with `drf-spectacular` and needed a bit of dark magic to make it work. If there's an easier way, let me know.

gergosimonyi commented on web/src/admin/policies/geoip/GeoIPPolicyForm.ts:
```diff
@@ -0,0 +1,170 @@
+                <div slot="body" class="pf-c-form">
+                    <ak-form-element-horizontal label=${msg("ASN Mode")} name="asnMode">
+                        <ak-search-select
+                            .fetchObjects=${async () => {
+                                return [
+                                    { name: "allow", label: "Allow" },
+                                    { name: "block", label: "Block" },
+                                ];
+                            }}
```
The use of an `async` function here for a static list of 2 is probably the worst part of this PR. I'll look to improve this.

kensternberg-authentik replied:
Yeah, `search select` is specifically for complex objects that have subtitles and complex interactions.  Consider using a Switch for a boolean choice, or if the ergonomics are bad (it's sometimes hard to word), a Radio.

gergosimonyi commented on web/src/admin/policies/geoip/GeoIPPolicyForm.ts:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    async send(data: GeoIPPolicy): Promise<GeoIPPolicy> {
+        if (data.asnMode?.toString() === "") {
+            data.asnMode = "block";
+        }
+
+        if (data.asnList?.toString() === "") {
+            data.asnList = [];
+        } else {
+            data.asnList = data.asnList.split(",").map(Number);
+        }
+
+        if (data.countryMode?.toString() === "") {
+            data.countryMode = "block";
+        }
+
+        if (data.countryList?.toString() === "") {
+            data.countryList = [];
+        } else {
+            data.countryList = data.countryList.split(",");
+        }
```
Defaults should be part of the back-end. Also some botched types here.

kensternberg-authentik replied:
Yeah, I'm looking at this and thinking that I'm seeing a lot of mixed messages compared to the types generated from the schema.  You shouldn't have to be splitting anything; the API should be giving those to you, and the TypeScript definitions in GeoIPPolicy.ts say getting strings where arrays are expected shouldn't be possible).  This function is probably the wrong place to be doing this.

If you're crafting a blob to `send`, start with a default and overwrite it with what you get from the form instead.

kensternberg-authentik commented on web/src/admin/policies/geoip/GeoIPPolicyForm.ts:
```diff
@@ -0,0 +1,170 @@
+                )}
+            </span>
+            <ak-form-element-horizontal label=${msg("Name")} ?required=${true} name="name">
+                <input
```
`?required=${true}` isn't needed here (and I _really_ need to break @beryju of the habit).  `required` is all that's needed.  You only need the `?` syntax for boolean attributes if the attribute can be changed by client code, for example: 
`?hidden=${!this.open}`

kensternberg-authentik commented on web/src/admin/policies/geoip/GeoIPPolicyForm.ts:
```diff
@@ -0,0 +1,170 @@
+                    <input
+                        class="pf-c-switch__input"
+                        type="checkbox"
+                        ?checked=${first(this.instance?.executionLogging, false)}
```
`first` is a leftover from earlier iterations of TypeScript.  Please use the idiom:
```${this.instance?.executionLogging ?? false}```

Although in this case I suspect you could just use the boolean forcing expression: `${!!this.instance?.executionLogging}`

kensternberg-authentik commented on web/src/admin/policies/geoip/GeoIPPolicyForm.ts:
```diff
@@ -0,0 +1,170 @@
+                        </p>
+                    </ak-form-element-horizontal>
+                    <ak-form-element-horizontal label=${msg("Country Mode")} name="countryMode">
+                        <ak-search-select
```
... and given that you've done it twice, break it out into a variable or something so you'll only have to change it once if the rules change.  :-)

kensternberg-authentik commented on web/src/admin/policies/geoip/GeoIPPolicyForm.ts:
```diff
@@ -0,0 +1,170 @@
+                    <ak-form-element-horizontal label=${msg("Country List")} name="countryList">
+                        <input
+                            type="text"
+                            value="${ifDefined(this.instance?.countryList || "")}"
```
Since you've provided a not-`undefined` value for falsity, `ifDefined(this.instance?.countryList || "")` will always be defined.  You can replace that whole expression with `this.instance?.countryList ?? ""`

kensternberg-authentik reviewed (commented):
Overall, the Typescript is really promising!  Good work.  There's a bit of rough edges, and it's obviously not done, but it's getting there.

gergosimonyi committed:
handle empty lists of ASNs and countries

gergosimonyi committed:
handle missing GeoIP database or missing IP from the database

The exceptions raised here are `PolicyException`s to let admins bypass
an execution failure.

gergosimonyi committed:
fix translations

whoops

BeryJu commented on authentik/policies/geoip/models.py:
```diff
@@ -0,0 +1,81 @@
+        asn = asn_data.get("asn") if asn_data else None
+        country = geoip_data.get("country") if geoip_data else None
+
+        if not asn:
```
both of these cases can also happen if ASN/GeoIP is not configured (pretty rare since we include the DB by default, but it is possible to disable)

gergosimonyi replied:
(We've had a bit of an offline discussion, but no resolution to this.) I kept the behavior for now, this way we don't leak information on whether a GeoIP database is available.

Let me know if you don't agree.

BeryJu commented on authentik/policies/geoip/models.py:
```diff
@@ -0,0 +1,81 @@
+from authentik.policies.types import PolicyRequest, PolicyResult
+
+
+class GeoIPPolicyMode(models.TextChoices):
```
Usually we don't have this as a setting directly on the policy but instead use the `negate` flag on the policy binding to inverse the effects. For this case I'm a bit on the fence as you can do two different things in the same policy with it

gergosimonyi replied:
Oh, I didn't notice that flag! That makes everything so much simpler. I'm not on the fence about it at all :smile: 

Then there's no need for `*_mode` and this model will just have 2 fields: `asn_list` and `country_list`. Same as Event Matcher Policy for least astonishment:

> If any of the configured values match, the policy passes.

BeryJu commented on web/src/admin/policies/geoip/GeoIPPolicyForm.ts:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    async send(data: GeoIPPolicy): Promise<GeoIPPolicy> {
+        if (data.asnMode?.toString() === "") {
```
This is somewhat of a downside of the generated API client, but this value here should never be undefined so this if shouldn't be required

gergosimonyi replied:
Combined with the check below (which is still needed because `"".split(",")` is non-empty)

BeryJu commented on web/src/admin/policies/geoip/GeoIPPolicyForm.ts:
```diff
@@ -0,0 +1,170 @@
+                        <ak-search-select
+                            .fetchObjects=${async () => {
+                                return [
+                                    { name: "allow", label: "Allow" },
```
the value for this (regardless of which UI component is used) should use the enum from the API instead of the raw string

gergosimonyi committed:
remove `GeoIPPolicyMode`

Use the policy binding's `negate` option instead

gergosimonyi committed:
use `django-countries` instead of a static list of countries for ISO-3166

gergosimonyi committed:
fix `DataProvision` typing

`ak-dual-select-provider` can handle unpaginated data

gergosimonyi committed:
simplify `GeoIPPolicyForm`

gergosimonyi committed:
Merge branch 'main' into policies/add-geoip-policy

gergosimonyi commented on authentik/policies/geoip/api.py:
```diff
@@ -0,0 +1,49 @@
+from authentik.policies.api.policies import PolicySerializer
+from authentik.policies.geoip.models import GeoIPPolicy
+
+COUNTRY_SCHEMA = {
+    "code": CountryField(),
+    "name": serializers.CharField(),
+}
```
This is an exact copy of `authentik/policies/geoip/views:11-14`. Weirdly, importing it from there results in `gen-build` silently failing to generate this part of the schema. I'm out of my depth here, so I just duplicated these 4 lines.

BeryJu replied:
That is very odd but also doesn't surprise me

gergosimonyi commented on authentik/policies/geoip/urls.py:
```diff
@@ -0,0 +1,11 @@
+
+api_urlpatterns = [
+    ("policies/geoip", GeoIPPolicyViewSet),
+    path("iso-3166/", ISO3166View.as_view(), name="iso-3166-view"),
```
There may or may not be a better place for this.

@BeryJu?

BeryJu replied:
I'd usually prefer to do it as `policies/geoip/iso3166/`, but obviously that could potentially cause errors due to that path also having the primary key...as a fallback I'd go with `policies/geoip_iso3166/` I guess?

gergosimonyi commented on web/src/admin/policies/geoip/GeoIPPolicyForm.ts:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    async send(data: GeoIPPolicy): Promise<GeoIPPolicy> {
+        if (data.asns?.toString() === "") {
+            data.asns = [];
+        } else {
+            data.asns = (data.asns as unknown as string).split(",").map(Number);
+        }
```
The actual type of `data` slightly differs from `GeoIPPolicy`.

I haven't found an easy way to reconcile them (the "proper" way is probably building a component on top of `<input>` to run a transforming function when its `value` is read), so I opted for the type assertion.

gergosimonyi commented on web/src/admin/policies/geoip/GeoIPPolicyForm.ts:
```diff
@@ -0,0 +1,125 @@
+                    </ak-form-element-horizontal>
+                    <ak-form-element-horizontal label=${msg("Countries")} name="countries">
+                        <ak-dual-select-provider
+                            .provider=${(page: number, search?: string): Promise<DataProvision> => {
+                                return new Iso3166Api(DEFAULT_CONFIG)
+                                    .iso3166List()
+                                    .then((results) => {
+                                        if (!search) return results;
+                                        return results.filter((result) =>
+                                            result.name
+                                                .toLowerCase()
+                                                .includes(search.toLowerCase()),
+                                        );
+                                    })
+                                    .then((results) => {
+                                        return {
+                                            options: results.map((country) => [
+                                                country.code,
+                                                country.name,
+                                            ]),
+                                        };
+                                    });
+                            }}
```
This currently hits the API for no reason every time there's a change in `search`. I'll add an optimization.

BeryJu reviewed (commented):
mostly LGTM

gergosimonyi committed:
pass `GeoIPPolicy` on empty policy

gergosimonyi committed:
add backend tests to `GeoIPPolicy`

gergosimonyi committed:
revise translations

gergosimonyi committed:
move `iso-3166/` to `policies/geoip_iso3166/`

gergosimonyi committed:
add client-side caching to ISO3166 API call

gergosimonyi committed:
fix `GeoIPPolicy` creation

The automatically generated APIs can't seem to handle `CountryField`,
so I'll have to do this by hand too.

gergosimonyi committed:
add docs for GeoIP Policy

tanberry commented on website/docs/policies/index.md:
```diff
@@ -10,6 +10,10 @@ This policy is used by the events subsystem. You can use this policy to match ev
 
 See [Expression Policy](expression.mdx).
 
+## GeoIP Policy
```
Hmmm I see that all of the other headers/titles for policies also capitalize "Policy". Our style is to use sentence style capitalization for headings... do you mind also changing all of the other headings to on this age to use a lower-case "p" on "policies"? :-)

tanberry reviewed (changes requested):
Yay we have a GeoIP policy! Thanks so much for updating the docs. A few nits, mostly about capitalization... I see that this whole page does not follow our Style Guide about using sentence-case capitalization for headers/titles. ("GeoIP policy" not "GeoIP Policy".)

gergosimonyi commented on website/docs/policies/expression.mdx:
````diff
@@ -55,18 +55,26 @@ import Objects from "../expressions/_objects.md";
 
     ```python
-    return context["geoip"]["country"] == "US"
+    return context["geoip"]["continent"] == "EU"
````
It's possible, e.g. countries in Eurovision famously include Australia :slightly_smiling_face: 

```
    return context["geoip"]["continent"] == "EU" or context["geoip"]["country"] == "AU"
```

but I believe this might be more confusing than just keeping it simple. The params are listed right above, so I don't think "new confusion" arises from changing this.

(I changed this just to suggesting something other that what could be done with a GeoIP policy.)

gergosimonyi committed:
docs: stylize

add review suggestions from @tanberry

tanberry reviewed (approved):
Thanks so much for updating the Docs!

gergosimonyi committed:
Merge branch 'main' into policies/add-geoip-policy

kensternberg-authentik reviewed (approved):
Overall, this looks quite good.  I have a few quibbles about syntax and idioms, but they're more of a "I wouldn't do it this way, but it works and it's not wrong, so let's go with it." If you want to tighten up a few of the wordier expressions, feel free, but otherwise I'm happy.

gergosimonyi committed:
refactor `GeoIPPolicy` API

It is now as declarative as I could make it.

gergosimonyi committed:
Merge branch 'main' into policies/add-geoip-policy

rissson commented on authentik/policies/geoip/models.py:
```diff
@@ -0,0 +1,92 @@
+        return result
+
+    def passes_asn(self, request: PolicyRequest) -> PolicyResult:
+        # This is not a single get chain because `request.context` can contain `{ "asn": None }`.
+        asn_data = request.context.get("asn")
+        asn = asn_data.get("asn") if asn_data else None
```
```suggestion
        asn = request.context.get("asn", {}).get("asn", None)
```
should work as well (the same applies for countries below)

gergosimonyi replied:
Unfortunately, no! If `request.context == { "asn": None }`, then `request.context.get("asn", {})` will happily return `None` and the second `.get()` will fail with

> AttributeError: 'NoneType' object has no attribute 'get'

rissson replied:
Correct, I didn't know you could get none in there

rissson commented on authentik/policies/geoip/views.py:
```diff
@@ -0,0 +1,20 @@
+    name = serializers.CharField()
+
+
+class ISO3166View(ListAPIView):
```
I think there is an argument for putting this in `api/` (or api.py) as well, for consistency with how we do things elsewhere.

gergosimonyi replied:
That makes sense. This used to be much more convoluted, but I should have moved it when I switched it to declarative style. Done!

rissson reviewed (approved):
Some nitpicking, but otherwise looks quite good!

gergosimonyi committed:
clean up `api.py` and `views.py`

gergosimonyi committed:
Merge branch 'main' into policies/add-geoip-policy

gergosimonyi committed:
Merge branch 'main' into policies/add-geoip-policy
